File: admin-ui/src/utils/currencyHelper.ts
import { CurrencyCode } from '../api/gamesApi';

export interface CurrencyInfo {
  code: CurrencyCode;
  name: string;
  symbol: string;
  fractionalUnit: string;
  minIncrement: number;
  requiresSpecialHandling: boolean;
  description?: string;
}

// Define all currency information
export const CURRENCY_INFO: Record<CurrencyCode, CurrencyInfo> = {
  [CurrencyCode.AED]: {
    code: CurrencyCode.AED,
    name: "United Arab Emirates Dirham",
    symbol: "د.إ",
    fractionalUnit: "fils",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.AUD]: {
    code: CurrencyCode.AUD,
    name: "Australian Dollar",
    symbol: "A$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.BRL]: {
    code: CurrencyCode.BRL,
    name: "Brazilian Real",
    symbol: "R$",
    fractionalUnit: "centavos",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.CAD]: {
    code: CurrencyCode.CAD,
    name: "Canadian Dollar",
    symbol: "C$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.CHF]: {
    code: CurrencyCode.CHF,
    name: "Swiss Franc",
    symbol: "Fr.",
    fractionalUnit: "centime",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.CLP]: {
    code: CurrencyCode.CLP,
    name: "Chilean Peso",
    symbol: "$",
    fractionalUnit: "centavos",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 centavos"
  },
  [CurrencyCode.CNY]: {
    code: CurrencyCode.CNY,
    name: "Chinese Renminbi",
    symbol: "¥",
    fractionalUnit: "fēn",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.COP]: {
    code: CurrencyCode.COP,
    name: "Colombian Peso",
    symbol: "$",
    fractionalUnit: "centavos",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 centavos"
  },
  [CurrencyCode.CRC]: {
    code: CurrencyCode.CRC,
    name: "Costa Rican Colón",
    symbol: "₡",
    fractionalUnit: "céntimos",
    minIncrement: 500,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 500 céntimos"
  },
  [CurrencyCode.EUR]: {
    code: CurrencyCode.EUR,
    name: "Euro",
    symbol: "€",
    fractionalUnit: "eurocents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.GBP]: {
    code: CurrencyCode.GBP,
    name: "British Pound",
    symbol: "£",
    fractionalUnit: "pence",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.HKD]: {
    code: CurrencyCode.HKD,
    name: "Hong Kong Dollar",
    symbol: "HK$",
    fractionalUnit: "sin",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.ILS]: {
    code: CurrencyCode.ILS,
    name: "Israeli New Shekel",
    symbol: "₪",
    fractionalUnit: "agorot",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.IDR]: {
    code: CurrencyCode.IDR,
    name: "Indonesian Rupiah",
    symbol: "Rp",
    fractionalUnit: "sen",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 sen"
  },
  [CurrencyCode.INR]: {
    code: CurrencyCode.INR,
    name: "Indian Rupee",
    symbol: "₹",
    fractionalUnit: "paise",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 paise"
  },
  [CurrencyCode.JPY]: {
    code: CurrencyCode.JPY,
    name: "Japanese Yen",
    symbol: "¥",
    fractionalUnit: "sen",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 sen"
  },
  [CurrencyCode.KRW]: {
    code: CurrencyCode.KRW,
    name: "South Korean Won",
    symbol: "₩",
    fractionalUnit: "jeon",
    minIncrement: 1000,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 1000 jeon"
  },
  [CurrencyCode.KWD]: {
    code: CurrencyCode.KWD,
    name: "Kuwaiti Dinar",
    symbol: "د.ك",
    fractionalUnit: "fils",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.KZT]: {
    code: CurrencyCode.KZT,
    name: "Kazakhstani Tenge",
    symbol: "₸",
    fractionalUnit: "tïın",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 tïın"
  },
  [CurrencyCode.MXN]: {
    code: CurrencyCode.MXN,
    name: "Mexican Peso",
    symbol: "Mex$",
    fractionalUnit: "centavos",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.MYR]: {
    code: CurrencyCode.MYR,
    name: "Malaysian Ringgit",
    symbol: "RM",
    fractionalUnit: "sen",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.NOK]: {
    code: CurrencyCode.NOK,
    name: "Norwegian Krone",
    symbol: "kr",
    fractionalUnit: "Øre",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.NZD]: {
    code: CurrencyCode.NZD,
    name: "New Zealand Dollar",
    symbol: "NZ$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.PEN]: {
    code: CurrencyCode.PEN,
    name: "Peruvian Sol",
    symbol: "S/",
    fractionalUnit: "céntimos",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.PHP]: {
    code: CurrencyCode.PHP,
    name: "Philippine Peso",
    symbol: "₱",
    fractionalUnit: "centavos",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.PLN]: {
    code: CurrencyCode.PLN,
    name: "Polish Złoty",
    symbol: "zł",
    fractionalUnit: "grosz",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.QAR]: {
    code: CurrencyCode.QAR,
    name: "Qatari Riyal",
    symbol: "ر.ق",
    fractionalUnit: "dirham",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.RUB]: {
    code: CurrencyCode.RUB,
    name: "Russian Ruble",
    symbol: "₽",
    fractionalUnit: "kopeks",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.SAR]: {
    code: CurrencyCode.SAR,
    name: "Saudi Riyal",
    symbol: "ر.س",
    fractionalUnit: "halalah",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.SGD]: {
    code: CurrencyCode.SGD,
    name: "Singapore Dollar",
    symbol: "S$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.THB]: {
    code: CurrencyCode.THB,
    name: "Thai Baht",
    symbol: "฿",
    fractionalUnit: "satang",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.TWD]: {
    code: CurrencyCode.TWD,
    name: "New Taiwan Dollar",
    symbol: "NT$",
    fractionalUnit: "fēn",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 fēn"
  },
  [CurrencyCode.UAH]: {
    code: CurrencyCode.UAH,
    name: "Ukrainian Hryvnia",
    symbol: "₴",
    fractionalUnit: "kopiykas",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 kopiykas"
  },
  [CurrencyCode.USD]: {
    code: CurrencyCode.USD,
    name: "United States Dollar",
    symbol: "$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  },
  [CurrencyCode.USD_CIS]: {
    code: CurrencyCode.USD_CIS,
    name: "US Dollar (CIS)",
    symbol: "$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false,
    description: "Discounted USD for Commonwealth of Independent States"
  },
  [CurrencyCode.USD_LATAM]: {
    code: CurrencyCode.USD_LATAM,
    name: "US Dollar (LATAM)",
    symbol: "$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false,
    description: "Discounted USD for Latin America"
  },
  [CurrencyCode.USD_MENA]: {
    code: CurrencyCode.USD_MENA,
    name: "US Dollar (MENA)",
    symbol: "$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false,
    description: "Discounted USD for Middle East and North Africa"
  },
  [CurrencyCode.USD_SASIA]: {
    code: CurrencyCode.USD_SASIA,
    name: "US Dollar (South Asia)",
    symbol: "$",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false,
    description: "Discounted USD for South Asia"
  },
  [CurrencyCode.UYU]: {
    code: CurrencyCode.UYU,
    name: "Uruguayan Peso",
    symbol: "$U",
    fractionalUnit: "centesimos",
    minIncrement: 100,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 100 centesimos"
  },
  [CurrencyCode.VND]: {
    code: CurrencyCode.VND,
    name: "Vietnamese Dong",
    symbol: "₫",
    fractionalUnit: "xu",
    minIncrement: 50000,
    requiresSpecialHandling: true,
    description: "Must be charged in increments of 50000 xu"
  },
  [CurrencyCode.ZAR]: {
    code: CurrencyCode.ZAR,
    name: "South African Rand",
    symbol: "R",
    fractionalUnit: "cents",
    minIncrement: 1,
    requiresSpecialHandling: false
  }
};

// Returns all currencies as an array for dropdown usage
export const getAllCurrencies = (): CurrencyInfo[] => {
  return Object.values(CURRENCY_INFO);
};

// Gets a specific currency's info
export const getCurrencyInfo = (code: CurrencyCode): CurrencyInfo => {
  return CURRENCY_INFO[code];
};

// Format a price amount according to the currency's conventions
export const formatPrice = (code: CurrencyCode, amount: number): string => {
  const currencyInfo = CURRENCY_INFO[code];
  
  // For currencies with special increments
  if (currencyInfo.minIncrement >= 100) {
    const mainUnit = Math.floor(amount / currencyInfo.minIncrement);
    return `${currencyInfo.symbol}${mainUnit}`;
  }
  
  // Standard format
  const mainUnit = Math.floor(amount / 100);
  const subUnit = amount % 100;
  
  if (subUnit === 0) {
    return `${currencyInfo.symbol}${mainUnit}`;
  } else {
    return `${currencyInfo.symbol}${mainUnit}.${subUnit.toString().padStart(2, '0')}`;
  }
};

// Validate a price amount meets the currency's constraints
export const validatePriceAmount = (code: CurrencyCode, amount: number): boolean => {
  const currencyInfo = CURRENCY_INFO[code];
  return amount % currencyInfo.minIncrement === 0;
};

// Get default currency settings for a currency
export const getDefaultCurrencySettings = (code: CurrencyCode): {
  code: CurrencyCode;
  min_price_increment: number;
  fractional_unit: string;
} => {
  const info = CURRENCY_INFO[code];
  return {
    code,
    min_price_increment: info.minIncrement,
    fractional_unit: info.fractionalUnit
  };
};

File: server/app/api/helpers/currency_helper.py
"""Helper functions for currency operations."""
from typing import Dict, Optional

# Currency constraints for Steam supported currencies
CURRENCY_CONSTRAINTS = {
    "AED": {"fraction_name": "fils", "min_increment": 1},
    "AUD": {"fraction_name": "cents", "min_increment": 1},
    "BRL": {"fraction_name": "centavos", "min_increment": 1},
    "CAD": {"fraction_name": "cents", "min_increment": 1},
    "CHF": {"fraction_name": "centime", "min_increment": 1},
    "CLP": {"fraction_name": "centavos", "min_increment": 100},  # Must be in increments of 100
    "CNY": {"fraction_name": "fēn", "min_increment": 1},
    "COP": {"fraction_name": "centavos", "min_increment": 100},  # Must be in increments of 100
    "CRC": {"fraction_name": "céntimos", "min_increment": 500},  # Must be in increments of 500
    "EUR": {"fraction_name": "eurocents", "min_increment": 1},
    "GBP": {"fraction_name": "pence", "min_increment": 1},
    "HKD": {"fraction_name": "sin", "min_increment": 1},
    "ILS": {"fraction_name": "agorot", "min_increment": 1},
    "IDR": {"fraction_name": "sen", "min_increment": 100},  # Must be in increments of 100
    "INR": {"fraction_name": "paise", "min_increment": 100},  # Must be in increments of 100
    "JPY": {"fraction_name": "sen", "min_increment": 100},  # Must be in increments of 100
    "KRW": {"fraction_name": "jeon", "min_increment": 1000},  # Must be in increments of 1000
    "KWD": {"fraction_name": "fils", "min_increment": 1},
    "KZT": {"fraction_name": "tïın", "min_increment": 100},  # Must be in increments of 100
    "MXN": {"fraction_name": "centavos", "min_increment": 1},
    "MYR": {"fraction_name": "sen", "min_increment": 1},
    "NOK": {"fraction_name": "Øre", "min_increment": 1},
    "NZD": {"fraction_name": "cents", "min_increment": 1},
    "PEN": {"fraction_name": "céntimos", "min_increment": 1},
    "PHP": {"fraction_name": "centavos", "min_increment": 1},
    "PLN": {"fraction_name": "grosz", "min_increment": 1},
    "QAR": {"fraction_name": "dirham", "min_increment": 1},
    "RUB": {"fraction_name": "kopeks", "min_increment": 1},
    "SAR": {"fraction_name": "halalah", "min_increment": 1},
    "SGD": {"fraction_name": "cents", "min_increment": 1},
    "THB": {"fraction_name": "satang", "min_increment": 1},
    "TWD": {"fraction_name": "fēn", "min_increment": 100},  # Must be in increments of 100
    "UAH": {"fraction_name": "kopiykas", "min_increment": 100},  # Must be in increments of 100
    "USD": {"fraction_name": "cents", "min_increment": 1},
    "USD_CIS": {"fraction_name": "cents", "min_increment": 1},
    "USD_LATAM": {"fraction_name": "cents", "min_increment": 1},
    "USD_MENA": {"fraction_name": "cents", "min_increment": 1},
    "USD_SASIA": {"fraction_name": "cents", "min_increment": 1},
    "UYU": {"fraction_name": "centesimos", "min_increment": 100},  # Must be in increments of 100
    "VND": {"fraction_name": "xu", "min_increment": 50000},  # Must be in increments of 50000
    "ZAR": {"fraction_name": "cents", "min_increment": 1},
}

def get_currency_constraints(currency_code: str) -> Optional[Dict]:
    """
    Get constraints for a specific currency.
    
    Args:
        currency_code: The currency code to get constraints for
        
    Returns:
        Dict with constraints or None if not found
    """
    return CURRENCY_CONSTRAINTS.get(currency_code)

def validate_price_amount(currency_code: str, amount: int) -> bool:
    """
    Validate that a price amount meets the constraints for a specific currency.
    
    Args:
        currency_code: The currency code
        amount: The price amount in smallest units (e.g., cents)
        
    Returns:
        True if valid, False otherwise
    """
    constraints = get_currency_constraints(currency_code)
    if not constraints:
        return False
    
    min_increment = constraints.get("min_increment", 1)
    return amount % min_increment == 0

def format_price(currency_code: str, amount: int) -> str:
    """
    Format a price amount according to the currency's conventions.
    
    Args:
        currency_code: The currency code
        amount: The price amount in smallest units (e.g., cents)
        
    Returns:
        Formatted price string
    """
    constraints = get_currency_constraints(currency_code)
    if not constraints:
        return f"{currency_code} {amount}"
    
    # For currencies like JPY where the smallest unit is the main unit
    if constraints.get("min_increment", 1) >= 100:
        main_unit = amount // constraints["min_increment"]
        return f"{currency_code} {main_unit}"
    
    # Standard format for most currencies
    main_unit = amount // 100
    sub_unit = amount % 100
    
    if sub_unit == 0:
        return f"{currency_code} {main_unit}"
    else:
        return f"{currency_code} {main_unit}.{sub_unit:02d}"